### DISCUSSION TITLE
Is there a way to get  numerical values from Typst locate or state  function calls?

### DISCUSSION THREAD

**Discussion by PEWOLO** [typst/typst#3330] [Q&A]
Hello beautiful and hard working people,

I was wondering if there is a possibility to turn Typst locate or state function calls to numerical values?
Any idea would be super helpful.

Thank you!

**Comment by Enivex** [comment]
No, you can only access the value inside locate. So do what you want with it inside the function call.

**Reply by PEWOLO** [reply]
I am making slides template and need the total page count to world with an indicator that tell you how far in your slide you are.
See image below!

<img width="312" alt="screen" src="https://github.com/typst/typst/assets/62015594/e4456540-fe7c-4868-ac16-606427959462">

**Reply by Enivex** [reply]
There's nothing to prevent you from doing that inside the locate() call

**Comment by PEWOLO** [comment]
I have a lot of other  functions that depend on this variable. 
So putting everything inside  a single function will become a bit messy and hard to keep track of.

**Reply by Enivex** [reply]
I would suggest you check out the context part of https://laurmaedje.github.io/posts/types-and-context/

What you're trying to get around is just fundamentally impossible.

That being said, I suspect you've designed your template in a way that makes life harder for yourself than it needs to be. You may want to post a question in the support channel on Discord. It's a bit easier than using GitHub, especially because of the bot.

**Reply by PEWOLO** [reply]
Thank you so much for your help!
I appreciated it

**Comment by Julian-Wassmann** [comment]
I've run into the same problem as well.

First of all, `locate()` will display the return value in the document. It is not directly possible to save its value as an type other than `content`.

In most cases it is the easiest to perform the necessary operations inside the `locate` function and display the result as `content`. This is what @Enivex suggested. It might be possible to restructure your code to do this. However, on complex use cases, this is not very practical.

**Edit:** The method shown below is not the best practice for this use case. Read the comments below for a better approach.

There is another trick: You can use a state to save the output of the `locate` call. This way, the type of the saved variable persits. Take a look at the following example. Note that to retrieve a state, there are two ways: Using `state.display()` and `state.at(loc)`. Only the last one will return the value with its corresponding type. You'll need to pass a location for evaluating the state.

```typst
= Heading
#let levelState = state("level", none)

#locate(loc => {
  let headings = query(heading, loc)
  if headings.len() == 0 {return}
  let level = headings.at(0).level
  levelState.update(level)
})

== Using state.display()
#{
  let level = levelState.display()
  [`level` is of type #type(level).]
}

== Using state.at(loc)
#locate(loc => {
  let level = levelState.at(loc)
  [`level` is of type #type(level).]
})
```

**Reply by PEWOLO** [reply]
Thank you so much Julian!
That was helpful

**Reply by laurmaedje** [reply]
What is the benefit of doing it like this? You still need to use locate to read the state after all. 

Generally, this will be less efficient (the compiler will run more internal iterations to resolve this). I'd be curious about use cases that this unlocks which are not possible in the other approach.

**Reply by Enivex** [reply]
This seems like an overly complicated way to go about it @Julian-Wassmann. It's much easier to define a function `level` which takes in `loc` and provides the value.

If you simply need to display it, then do `locate(level)`. Otherwise you would do `locate(loc => do something with level(loc))`

**Reply by Julian-Wassmann** [reply]
Okay, I gave some thoughts about this. I'm all agreeing to your objection now. I can't imagine about a case right now, where you couldn't use a function in a `locate` call. Thank you for your input!

I'm still gonna leave my anwser, because it's a way which is safe to use. But I will add a note for clarification.
